src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -4782,8 +4782,33 @@ void MacroAssembler::string_equals(Register a1, Register a2,
 
   mov(result, false);
 
+  // Check for short strings, i.e. smaller than wordSize * 2.
+  subs(cnt1, cnt1, wordSize * 2);
+  br(Assembler::LT, MEDIUM);
+  // Main 16 byte comparison loop.
+  bind(NEXT_TWO_WORDS); {
+    ldr(v0, Q, Address(post(a1, wordSize * 2)));
+    ldr(v1, Q, Address(post(a2, wordSize * 2)));
+    subs(cnt1, cnt1, wordSize * 2);
+    eor(v0, T16B, v0, v1);
+    mov(tmp1, v0, T2D, 0);
+    cbnz(tmp1, DONE);
+    mov(tmp2, v0, T2D, 1);
+    cbnz(tmp2, DONE);
+  } br(GT, NEXT_TWO_WORDS);
+  // Same as MEDIUM
+  ldr(v0, Q, Address(a1, cnt1));
+  ldr(v1, Q, Address(a2, cnt1));
+  eor(v0, T16B, v0, v1);
+  mov(tmp1, v0, T2D, 0);
+  cbnz(tmp1, DONE);
+  mov(tmp2, v0, T2D, 1);
+  cbnz(tmp2, DONE);
+  b(SAME);
```

Comment:
Shouldn't this be
```
mov(tmp1, v0, T2D, 0);
mov(tmp2, v0, T2D, 1);
orr(tmp1, tmp1, tmp2);
cbnz(tmp1, DONE);
```

... which would use up fewer branch prediction resources.

Reply:
... or maybe do the OR in the vector unit?

Reply:
Sure, great idea.

Reply:
I have tested @dgbo 's suggestion and found that the performance degradation happened by using `umaxv`.

Reply:
I guess I'm not surprised it's slower: even Firestorm has a 3-cycle latency for UMAX, and its output is used immediately.